src/Client/LocalConnection.cpp

```diff
@@ -0,0 +1,411 @@
+#include "LocalConnection.h"
+#include <Interpreters/executeQuery.h>
+#include <Storages/IStorage.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_PACKET_FROM_SERVER;
+    extern const int UNKNOWN_EXCEPTION;
+    extern const int NOT_IMPLEMENTED;
+}
+
+LocalConnection::LocalConnection(ContextPtr context_, bool send_progress_)
+    : WithContext(context_)
+    , session(getContext(), ClientInfo::Interface::LOCAL)
+    , send_progress(send_progress_)
+{
+    /// Authenticate and create a context to execute queries.
+    session.authenticate("default", "", Poco::Net::SocketAddress{});
+
+    if (!CurrentThread::isInitialized())
+        thread_status.emplace();
+
+    query_context = session.makeQueryContext();
+    query_context->makeSessionContext(); /// initial_create_query requires a session context to be set.
+    query_context->setCurrentQueryId("");
```

Comment:
Do we execute all queries with one `query_context` or do we recreate `LocalConnection` for each query?

Reply:
We create LocalConnection only once for the lifetime of LocalServer.

Reply:
We should create a session context for connection and separate query context for each query within the session

Reply:
Moving `query_context->setCurrentQueryId("");` from constructor to `sendQuery` is a dirty hack that fixes the specific issue with query id, but the root cause (and, probably, many similar issues) remains.

Reply:
> query context for each query within the session

ok, but should we follow this for non-interactive:
https://github.com/ClickHouse/ClickHouse/blob/64613913872afe82b421fe1e0c56bc90efbd1f8b/programs/local/LocalServer.cpp#L398

Reply:
Hmm, I don't know, the comment does not explain why we should use single context. Let's take a look through the history of changes:
https://github.com/ClickHouse/ClickHouse/blob/922cf7ee20200cc8564f0e6190f1d2da8900140c/programs/local/LocalServer.cpp#L394-L399

Before `Use sessions more #27817`:
https://github.com/ClickHouse/ClickHouse/blob/65ee9a1272725dd087437c16f591053b981269af/programs/local/LocalServer.cpp#L377-L382

Before `Add setTemporaryStorage to clickhouse-local to make OPTIMIZE work #16192`:
https://github.com/ClickHouse/ClickHouse/blob/9ffa7bd0b6812c7ca379da53133b2d4033bb1d01/programs/local/LocalServer.cpp#L338-L339

Global context was used everywhere in `LocalServer` (and it was incorrect), then some bug was fixed by copying the context. And then it just remained that way after some refactorings. Seems like we can change it.